Avoid unnecessary conversions from ASCIILiteral

brentfulgham opened the pull request:
#### 78bc243d1e64b30d8a4362fdacf666bffdc33e90
<pre>
Avoid unnecessary conversions from ASCIILiteral
<a href="https://bugs.webkit.org/show_bug.cgi?id=282724">https://bugs.webkit.org/show_bug.cgi?id=282724</a>
&lt;<a href="https://rdar.apple.com/problem/139394432">rdar://problem/139394432</a>&gt;

Reviewed by Chris Dumez.

Over many years of refactoring we&apos;ve wrapped immutable ASCIILiterals in
strings at various points. We should try to pass the basic type when
possible to avoid constructing a String object unless it is needed.

* Source/WTF/wtf/Hasher.h:
(WTF::add):
* Source/WebCore/bindings/js/CachedScriptFetcher.cpp:
(WebCore::CachedScriptFetcher::create):
* Source/WebCore/bindings/js/CachedScriptFetcher.h:
(WebCore::CachedScriptFetcher::CachedScriptFetcher):
* Source/WebCore/css/StyleRuleImport.cpp:
(WebCore::StyleRuleImport::requestStyleSheet):
(WebCore::StyleRuleImport::setCSSStyleSheet):
* Source/WebCore/css/StyleRuleImport.h:
* Source/WebCore/css/StyleSheetContents.h:
* Source/WebCore/css/parser/CSSParserContext.cpp:
(WebCore::CSSParserContext::CSSParserContext):
* Source/WebCore/css/parser/CSSParserContext.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::characterSetWithUTF8Fallback const):
(WebCore::Document::cssParserContext const):
(WebCore::Document::encoding const): Deleted.
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/DocumentInlines.h:
(WebCore::Document::encoding const):
(WebCore::Document::charset const):
* Source/WebCore/dom/InlineClassicScript.cpp:
(WebCore::InlineClassicScript::InlineClassicScript):
* Source/WebCore/dom/InlineClassicScript.h:
* Source/WebCore/dom/LoadableClassicScript.cpp:
(WebCore::LoadableNonModuleScriptBase::LoadableNonModuleScriptBase):
(WebCore::LoadableClassicScript::create):
(WebCore::LoadableClassicScript::LoadableClassicScript):
* Source/WebCore/dom/LoadableClassicScript.h:
* Source/WebCore/dom/LoadableModuleScript.cpp:
(WebCore::LoadableModuleScript::create):
(WebCore::LoadableModuleScript::LoadableModuleScript):
* Source/WebCore/dom/LoadableModuleScript.h:
* Source/WebCore/dom/LoadableScript.cpp:
(WebCore::LoadableScript::LoadableScript):
* Source/WebCore/dom/LoadableScript.h:
* Source/WebCore/dom/ProcessingInstruction.cpp:
(WebCore::ProcessingInstruction::checkStyleSheet):
(WebCore::ProcessingInstruction::setCSSStyleSheet):
* Source/WebCore/dom/ProcessingInstruction.h:
* Source/WebCore/dom/ScriptElement.cpp:
(WebCore::ScriptElement::prepareScript):
* Source/WebCore/dom/ScriptElement.h:
(WebCore::ScriptElement::scriptCharset const):
* Source/WebCore/dom/ScriptElementCachedScriptFetcher.h:
(WebCore::ScriptElementCachedScriptFetcher::ScriptElementCachedScriptFetcher):
* Source/WebCore/dom/TextDecoder.cpp:
(WebCore::TextDecoder::encoding const):
* Source/WebCore/dom/TextEncoder.cpp:
(WebCore::TextEncoder::encoding const):
* Source/WebCore/dom/TextEncoder.h:
 * Source/WebCore/html/HTMLDetailsElement.cpp:
(WebCore::HTMLDetailsElement::otherElementsInNameGroup): Drive-by fix. Avoid doing lookup
Multiple times during loop.
* Source/WebCore/html/HTMLLinkElement.cpp:
(WebCore::HTMLLinkElement::setCSSStyleSheet):
* Source/WebCore/html/HTMLLinkElement.h:
* Source/WebCore/html/HTMLScriptElement.cpp:
(WebCore::HTMLScriptElement::charsetAttributeValue const): Drive-by fix. Avoid
Converting to string, when its only use cases are to copy back to an AtomString.
* Source/WebCore/html/HTMLScriptElement.h:
* Source/WebCore/loader/LinkPreloadResourceClients.h:
* Source/WebCore/loader/cache/CachedApplicationManifest.cpp:
(WebCore::CachedApplicationManifest::encoding const):
* Source/WebCore/loader/cache/CachedApplicationManifest.h:
* Source/WebCore/loader/cache/CachedCSSStyleSheet.cpp:
(WebCore::CachedCSSStyleSheet::didAddClient):
(WebCore::CachedCSSStyleSheet::encoding const):
(WebCore::CachedCSSStyleSheet::checkNotify):
* Source/WebCore/loader/cache/CachedCSSStyleSheet.h:
* Source/WebCore/loader/cache/CachedResource.h:
(WebCore::CachedResource::encoding const):
* Source/WebCore/svg/SVGScriptElement.h:
* Source/WebCore/loader/LinkPreloadResourceClients.h:
* Source/WebCore/loader/cache/CachedApplicationManifest.cpp:
(WebCore::CachedApplicationManifest::encoding const):
* Source/WebCore/loader/cache/CachedApplicationManifest.h:
* Source/WebCore/loader/cache/CachedCSSStyleSheet.cpp:
(WebCore::CachedCSSStyleSheet::didAddClient):
(WebCore::CachedCSSStyleSheet::encoding const):
(WebCore::CachedCSSStyleSheet::checkNotify):
* Source/WebCore/loader/cache/CachedCSSStyleSheet.h:
* Source/WebCore/loader/cache/CachedResource.h:
(WebCore::CachedResource::encoding const):
* Source/WebCore/loader/cache/CachedSVGDocument.cpp:
(WebCore::CachedSVGDocument::encoding const):
* Source/WebCore/loader/cache/CachedSVGDocument.h:
* Source/WebCore/loader/cache/CachedScript.cpp:
(WebCore::CachedScript::encoding const):
* Source/WebCore/loader/cache/CachedScript.h:
* Source/WebCore/loader/cache/CachedStyleSheetClient.h:
(WebCore::CachedStyleSheetClient::setCSSStyleSheet):
* Source/WebCore/loader/cache/CachedXSLStyleSheet.cpp:
(WebCore::CachedXSLStyleSheet::encoding const):
* Source/WebCore/loader/cache/CachedXSLStyleSheet.h:
* Source/WebCore/svg/SVGScriptElement.h:
* Source/WebKitLegacy/mac/DOM/DOMDocument.mm:
(-[DOMDocument inputEncoding]):
(-[DOMDocument charset]):
(-[DOMDocument characterSet]):

Canonical link: <a href="https://commits.webkit.org/286364@main">https://commits.webkit.org/286364@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9c3d6a4133d69923d46ca9d743402551d20878f8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/75719 "Passed style check") | [⏳ 🛠 ios ](https://ews-build.webkit.org/#/builders/iOS-17-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/28570 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/80207 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/26983 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/77835 "Passed tests") | [⏳ 🛠 ios-sim ](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/3007 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/59400 "Build is in progress. Recent messages:Printed configuration") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/17567 "Build is in progress. Recent messages:Printed configuration") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/78786 "Passed tests") | [⏳ 🧪 ios-wk2 ](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/65046 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39753 "Passed tests") | 
| | [⏳ 🧪 ios-wk2-wpt ](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/22525 "Passed tests") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/25312 "Build was cancelled. Recent messages:Printed configuration; Running apply-patch; Checked out pull request; compiling") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/68869 "Built successfully and passed tests") | [⏳ 🧪 api-ios ](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/22861 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/81678 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/74981 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/3058 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/1934 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/67629 "Build is in progress. Recent messages:Printed configuration") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/3209 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/65014 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/66929 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/10877 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/9014 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/97249 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11705 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/3015 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/5837 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/21260 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/3040 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/3975 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/3047 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/css/parser/CSSParserContext.h:
```diff
@@ -109,7 +109,7 @@ struct CSSParserContext {
     CSSParserContext(CSSParserMode, const URL& baseURL = URL());
     WEBCORE_EXPORT CSSParserContext(const Document&);
-    CSSParserContext(const Document&, const URL& baseURL, const String& charset = emptyString());
+    CSSParserContext(const Document&, const URL& baseURL, ASCIILiteral charset = ASCIILiteral());
```
I don't know if it matters but this is not equivalent to the old code. The equivalent would be `ASCIILiteral charset = ""_s`

brentfulgham replied:
Ah! Good catch. I will fix that. Maybe that is the source of my test failures.

cdumez commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -10407,7 +10401,7 @@ CSSCounterStyleRegistry& Document::counterStyleRegistry()
 {
     if (!m_cachedCSSParserContext)
-        m_cachedCSSParserContext = makeUnique<CSSParserContext>(*this, URL { }, emptyString());
+        m_cachedCSSParserContext = makeUnique<CSSParserContext>(*this, URL { }, ASCIILiteral());
```
Ditto.

brentfulgham commented on Source/WebCore/css/parser/CSSParserContext.cpp:
```diff
@@ -146,7 +146,7 @@ void add(Hasher& hasher, const CSSParserContext& context)
         | context.cssProgressFunctionEnabled                << 25
         | (uint32_t)context.mode                            << 26; // This is multiple bits, so keep it last.
-    add(hasher, context.baseURL, context.charset, context.propertySettings, bits);
+    add(hasher, context.baseURL, AtomString(context.charset), context.propertySettings, bits);
```
@cdumez Thoughts on this? You use a bespoke hasher in TextEncodingRegistry, which maybe should be part of Hasher.h so it could just accept ASCIILiteral.

I though AtomString might be good here, since it would presumably hash to a consistent value and would amortize the cost of casting to string since these encoding are always one of a small set.

cdumez replied:
We should be able to add this to Hasher.h or ASCIILiteral.h so that we can just pass an ASCIILiteral in:
```
inline void add(Hasher& hasher, ASCIILiteral literal)
{
    // Chose to hash the characters here. Assuming this is better than hashing the possibly-already-computed hash of the characters.
    bool remainder = literal.length() & 1;
    unsigned roundedLength = literal.length() - remainder;
    for (unsigned i = 0; i < roundedLength; i += 2)
        add(hasher, (literal[i] << 16) | literal[i + 1]);
    if (remainder)
        add(hasher, literal[roundedLength]);
}
```

brentfulgham replied:
Great! I will add that.

cdumez commented on Source/WebCore/css/StyleRuleImport.cpp:
```diff
@@ -135,7 +135,7 @@ void StyleRuleImport::requestStyleSheet()
     // FIXME: Skip Content Security Policy check when stylesheet is in a user agent shadow tree.
     // See <https://bugs.webkit.org/show_bug.cgi?id=146663>.
-    CachedResourceRequest request(absURL, CachedResourceLoader::defaultCachedResourceOptions(), std::nullopt, String(m_parentStyleSheet->charset()));
+    CachedResourceRequest request(absURL, CachedResourceLoader::defaultCachedResourceOptions(), std::nullopt, String::fromLatin1(m_parentStyleSheet->charset()));
```
Why do we need `::fromLatin1()` here? There is a `String(ASCIILiteral)` constructor available, which would be more efficient.

brentfulgham replied:
I used the `::fromLatin1()` because I saw it used elsewhere. I thought that since we know the `ASCIILiteral` is guaranteed to be 8-bit and Latin1, that it would be able to avoid some checks. But clearly not!

cdumez commented on Source/WebCore/dom/ScriptElement.h:
```diff
@@ -52,7 +52,7 @@ class ScriptElement {
     bool prepareScript(const TextPosition& scriptStartPosition = TextPosition());
 
-    String scriptCharset() const { return m_characterEncoding; }
+    AtomString scriptCharset() const { return m_characterEncoding; }
```
Ideally would return `const AtomString&` to not force the caller to copy.

brentfulgham replied:
Oh, good point. I think that's how it's done elsewhere, too.

cdumez commented on Source/WebCore/dom/TextDecoder.cpp:
```diff
@@ -83,7 +83,7 @@ ExceptionOr<String> TextDecoder::decode(std::optional<BufferSource::VariantType>
 String TextDecoder::encoding() const
 {
-    return makeString(asASCIILowercase(StringView::fromLatin1(m_textEncoding.name())));
+    return StringView::fromLatin1(m_textEncoding.name()).convertToASCIILowercase();
```
Do we really need the `::fromLatin1()` here? There is a `StringView(ASCIILiteral)` constructor that should be more efficient.

brentfulgham replied:
Will fix.

cdumez commented on Source/WebCore/dom/TextEncoder.cpp:
```diff
@@ -34,7 +34,7 @@ WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN
 namespace WebCore {
 
-String TextEncoder::encoding() const
+ASCIILiteral TextEncoder::encoding() const
```
nit: maybe this can be inlined in the header.

cdumez reviewed (approved):
LGTM with nits.

brentfulgham committed:
Avoid unnecessary conversions from ASCIILiteral
https://bugs.webkit.org/show_bug.cgi?id=282724
<rdar://problem/139394432>

Reviewed by Chris Dumez.

Over many years of refactoring we've wrapped immutable ASCIILiterals in
strings at various points. We should try to pass the basic type when
possible to avoid constructing a String object unless it is needed.

* Source/WTF/wtf/Hasher.h:
(WTF::add):
* Source/WebCore/bindings/js/CachedScriptFetcher.cpp:
(WebCore::CachedScriptFetcher::create):
* Source/WebCore/bindings/js/CachedScriptFetcher.h:
(WebCore::CachedScriptFetcher::CachedScriptFetcher):
* Source/WebCore/css/StyleRuleImport.cpp:
(WebCore::StyleRuleImport::requestStyleSheet):
(WebCore::StyleRuleImport::setCSSStyleSheet):
* Source/WebCore/css/StyleRuleImport.h:
* Source/WebCore/css/StyleSheetContents.h:
* Source/WebCore/css/parser/CSSParserContext.cpp:
(WebCore::CSSParserContext::CSSParserContext):
* Source/WebCore/css/parser/CSSParserContext.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::characterSetWithUTF8Fallback const):
(WebCore::Document::cssParserContext const):
(WebCore::Document::encoding const): Deleted.
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/DocumentInlines.h:
(WebCore::Document::encoding const):
(WebCore::Document::charset const):
* Source/WebCore/dom/InlineClassicScript.cpp:
(WebCore::InlineClassicScript::InlineClassicScript):
* Source/WebCore/dom/InlineClassicScript.h:
* Source/WebCore/dom/LoadableClassicScript.cpp:
(WebCore::LoadableNonModuleScriptBase::LoadableNonModuleScriptBase):
(WebCore::LoadableClassicScript::create):
(WebCore::LoadableClassicScript::LoadableClassicScript):
* Source/WebCore/dom/LoadableClassicScript.h:
* Source/WebCore/dom/LoadableModuleScript.cpp:
(WebCore::LoadableModuleScript::create):
(WebCore::LoadableModuleScript::LoadableModuleScript):
* Source/WebCore/dom/LoadableModuleScript.h:
* Source/WebCore/dom/LoadableScript.cpp:
(WebCore::LoadableScript::LoadableScript):
* Source/WebCore/dom/LoadableScript.h:
* Source/WebCore/dom/ProcessingInstruction.cpp:
(WebCore::ProcessingInstruction::checkStyleSheet):
(WebCore::ProcessingInstruction::setCSSStyleSheet):
* Source/WebCore/dom/ProcessingInstruction.h:
* Source/WebCore/dom/ScriptElement.cpp:
(WebCore::ScriptElement::prepareScript):
* Source/WebCore/dom/ScriptElement.h:
(WebCore::ScriptElement::scriptCharset const):
* Source/WebCore/dom/ScriptElementCachedScriptFetcher.h:
(WebCore::ScriptElementCachedScriptFetcher::ScriptElementCachedScriptFetcher):
* Source/WebCore/dom/TextDecoder.cpp:
(WebCore::TextDecoder::encoding const):
* Source/WebCore/dom/TextEncoder.cpp:
(WebCore::TextEncoder::encoding const):
* Source/WebCore/dom/TextEncoder.h:
 * Source/WebCore/html/HTMLDetailsElement.cpp:
(WebCore::HTMLDetailsElement::otherElementsInNameGroup): Drive-by fix. Avoid doing lookup
Multiple times during loop.
* Source/WebCore/html/HTMLLinkElement.cpp:
(WebCore::HTMLLinkElement::setCSSStyleSheet):
* Source/WebCore/html/HTMLLinkElement.h:
* Source/WebCore/html/HTMLScriptElement.cpp:
(WebCore::HTMLScriptElement::charsetAttributeValue const): Drive-by fix. Avoid
Converting to string, when its only use cases are to copy back to an AtomString.
* Source/WebCore/html/HTMLScriptElement.h:
* Source/WebCore/loader/LinkPreloadResourceClients.h:
* Source/WebCore/loader/cache/CachedApplicationManifest.cpp:
(WebCore::CachedApplicationManifest::encoding const):
* Source/WebCore/loader/cache/CachedApplicationManifest.h:
* Source/WebCore/loader/cache/CachedCSSStyleSheet.cpp:
(WebCore::CachedCSSStyleSheet::didAddClient):
(WebCore::CachedCSSStyleSheet::encoding const):
(WebCore::CachedCSSStyleSheet::checkNotify):
* Source/WebCore/loader/cache/CachedCSSStyleSheet.h:
* Source/WebCore/loader/cache/CachedResource.h:
(WebCore::CachedResource::encoding const):
* Source/WebCore/svg/SVGScriptElement.h:
* Source/WebCore/loader/LinkPreloadResourceClients.h:
* Source/WebCore/loader/cache/CachedApplicationManifest.cpp:
(WebCore::CachedApplicationManifest::encoding const):
* Source/WebCore/loader/cache/CachedApplicationManifest.h:
* Source/WebCore/loader/cache/CachedCSSStyleSheet.cpp:
(WebCore::CachedCSSStyleSheet::didAddClient):
(WebCore::CachedCSSStyleSheet::encoding const):
(WebCore::CachedCSSStyleSheet::checkNotify):
* Source/WebCore/loader/cache/CachedCSSStyleSheet.h:
* Source/WebCore/loader/cache/CachedResource.h:
(WebCore::CachedResource::encoding const):
* Source/WebCore/loader/cache/CachedSVGDocument.cpp:
(WebCore::CachedSVGDocument::encoding const):
* Source/WebCore/loader/cache/CachedSVGDocument.h:
* Source/WebCore/loader/cache/CachedScript.cpp:
(WebCore::CachedScript::encoding const):
* Source/WebCore/loader/cache/CachedScript.h:
* Source/WebCore/loader/cache/CachedStyleSheetClient.h:
(WebCore::CachedStyleSheetClient::setCSSStyleSheet):
* Source/WebCore/loader/cache/CachedXSLStyleSheet.cpp:
(WebCore::CachedXSLStyleSheet::encoding const):
* Source/WebCore/loader/cache/CachedXSLStyleSheet.h:
* Source/WebCore/svg/SVGScriptElement.h:
* Source/WebKitLegacy/mac/DOM/DOMDocument.mm:
(-[DOMDocument inputEncoding]):
(-[DOMDocument charset]):
(-[DOMDocument characterSet]):

Canonical link: https://commits.webkit.org/286364@main
